token-lending/js/src/state/obligation.ts

```diff
@@ -0,0 +1,132 @@
+    depositReserve: PublicKey;
+    depositedAmount: BN;
+    marketValue: BN; // decimals
+}
+
+export interface ObligationLiquidity {
+    borrowReserve: PublicKey;
+    cumulativeBorrowRateWads: BN; // decimals
+    borrowedAmountWads: BN; // decimals
+    marketValue: BN; // decimals
+}
+
+/** @internal */
+export interface ObligationDataFlat {
+    version: number;
+    lastUpdate: LastUpdate;
+    lendingMarket: PublicKey;
+    owner: PublicKey;
+    depositedValue: BN; // decimals
+    borrowedValue: BN; // decimals
+    allowedBorrowValue: BN; // decimals
+    unhealthyBorrowValue: BN; // decimals
+    depositsLen: number;
+    borrowsLen: number;
+    dataFlat: Buffer;
+}
+
+/** @internal */
+export const ObligationCollateralLayout = struct<ObligationCollateral>([
+    publicKey('depositReserve'),
+    u64('depositedAmount'),
+    u128('marketValue'),
+]);
+
+/** @internal */
+export const ObligationLiquidityLayout = struct<ObligationLiquidity>([
+    publicKey('borrowReserve'),
+    u128('cumulativeBorrowRateWads'),
+    u128('borrowedAmountWads'),
+    u128('marketValue'),
+]);
+
+/** @internal */
+export const ObligationLayout = struct<ObligationDataFlat>([
+    u8('version'),
+
+    LastUpdateLayout,
+
+    publicKey('lendingMarket'),
+    publicKey('owner'),
+    u128('depositedValue'),
+    u128('borrowedValue'),
+    u128('allowedBorrowValue'),
+    u128('unhealthyBorrowValue'),
+
+    u8('depositsLen'),
+    u8('borrowsLen'),
+    blob(ObligationCollateralLayout.span + 9 * ObligationLiquidityLayout.span, 'dataFlat'),
+]);
+
+export const isObligation = (info: AccountInfo<Buffer>): boolean => {
+    return info.data.length === ObligationLayout.span;
+};
+
+export const parseObligation: Parser<Obligation> = (pubkey: PublicKey, info: AccountInfo<Buffer>) => {
+    if (!isObligation(info)) return;
+
+    const buffer = Buffer.from(info.data);
+    const {
+        version,
+        lastUpdate,
+        lendingMarket,
+        owner,
+        depositedValue,
+        borrowedValue,
+        allowedBorrowValue,
+        unhealthyBorrowValue,
+        depositsLen,
+        borrowsLen,
+        dataFlat,
+    } = ObligationLayout.decode(buffer);
+
+    if (!version) return;
+
+    const depositsSpan = depositsLen * ObligationCollateralLayout.span;
+    const borrowsSpan = borrowsLen * ObligationLiquidityLayout.span;
+
+    const depositsBuffer = dataFlat.slice(0, depositsSpan);
+    const deposits = seq(ObligationCollateralLayout, depositsLen).decode(depositsBuffer);
+
+    const borrowsBuffer = dataFlat.slice(depositsSpan, depositsSpan + borrowsSpan);
+    const borrows = seq(ObligationLiquidityLayout, borrowsLen).decode(borrowsBuffer);
+
+    const obligation = {
+        version,
+        lastUpdate,
+        lendingMarket,
+        owner,
+        depositedValue,
+        borrowedValue,
+        allowedBorrowValue,
+        unhealthyBorrowValue,
+        deposits,
+        borrows,
+    } as Obligation;
```

Comment:
nit: this could break easily in the future, is it necessary? Could you do `const obligation: Obligation = ...`

Reply:
Changed to `const obligation: Obligation`, though it probably isn't necessary anyway as the function declares the return type.